### ISSUE TITLE
FTP server: upload should support async close

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3462]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)| @warner reported|
|-|-|
|Trac ID|trac#3462|
|Type|enhancement|
|Created|2008-10-07 01:48:11Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-async-close-3462|

As discussed [on the mailing list](http://twistedmatrix.com/pipermail/twisted-python/2008-October/018476.html), the FTP server in current trunk (r24958) does not have a mechanism for handling uploads that are not complete the moment the last byte is accepted by the consumer. This makes it difficult, for example, to implement an FTP frontend on a virtual filesystem that takes some amount of time/work to finish the upload process.

The proposed solution is to enhance [twisted/protocols/ftp.py#L1424 ftp:IWriteFile](twisted/protocols/ftp.py#L1424 ftp:IWriteFile) to add a ```close()``` method, which will be called after the consumer has been disconnected, but before the '226 Transfer Complete' response is sent. ```close()``` can return a Deferred, and the response will not be sent until after it fires. The value produced by this Deferred is ignored.

Other proposed solutions were:
 * modify ```unregisterProducer``` to allow it to return a Deferred
 * do something at the VFS level

A patch to implement this change is attached. It does not yet have specific test coverage, but of course the existing tests still pass.

Attachments:

* [t3462.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ec/6ec645b7e53b0ab491ade2e56342cc50d5209384/c3e07ed92a7e52f0422d08a2fe68f4fdf24a9c2b.diff) (1852 bytes) - added by warner on 2008-10-07 01:49:21Z - patch to implement the interface+implementation change
* [t3462-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ec/6ec645b7e53b0ab491ade2e56342cc50d5209384/e0ee6bd66edf3d84baa51f684bdd53215d2e9718.diff) (5312 bytes) - added by warner on 2010-02-22 21:57:55Z - new patch, adds tests
* [t3462-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ec/6ec645b7e53b0ab491ade2e56342cc50d5209384/b6ade2b2324493d59a4c5cdb29cd286dc1d72168.diff) (5861 bytes) - added by warner on 2010-02-23 02:48:19Z - new patch incorporating spiv's recommendations

<details><summary>Searchable metadata</summary>

```
trac-id__3462 3462
type__enhancement enhancement
reporter__warner warner
priority__highest highest
milestone__ 
branch__branches_ftp_async_close_3462 branches/ftp-async-close-3462
branch_author__warner warner
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1223344091000000 1223344091000000
changetime__1266929545000000 1266929545000000
version__None None
owner__ 
cc__exarkun cc__terrycojones cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @warner|
|-|-|

Seems like a good start.  Can you add unit tests?

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_2"></a>|@warner commented|
|-|-|

ok, new patch uploaded with unit tests. exarkun gave it a quick once-over and didn't scream.

**Comment by twisted-trac** [comment]
|<img alt="terrycojones's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|terrycojones commented|
|-|-|

I applied the latest path, examined the diff and the patched code, didn't see anything to object to. Ran trial twisted without problem.

Thanks for the beer, Brian.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|spiv commented|
|-|-|

I like this idea, and this implementation of it.  Just a couple of small nits:

 1. Given that you're reformatting the test_write docstring anyway, fix the typo of "with fires" (should be "which fires").
 1. I think docstrings are preferred to comments for test descriptions (even though test_ftp is pretty confused on this point).  So change the initial comment of `FTPCloseTest.test_write` to a docstring.  
 1. (Optional: perhaps a docstring on `FTPCloseTest` itself would be a good idea too, something like "Tests that the server invokes IWriteFile.close.")
 1. This needs a news file.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|spiv commented|
|-|-|

Looks like warner forgot to reinstate the 'review' keyword.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|spiv commented|
|-|-|

Looks ok to me.  I'll put it on a branch and see what the buildbot thinks.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|spiv commented|
|-|-|

(In [28450]) Branching to ftp-async-close-3462.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|spiv set status to `closed`|
|-|-|

(In [28453]) Merge ftp-async-close-3462

Author: warner
Reviewer: spiv
Fixes: [#3462](https://github.com/twisted/twisted/issues/3462)

twisted.protocol.ftp.IWriteFile now has a close() method, which can return a
Deferred. Previously a STOR command would finish immediately upon the receipt
of the last byte of the uploaded file. With close(), the backend can delay
the finish until it has performed some other slow action (like storing the
data to a virtual filesystem).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
